Hw1 add bot skeleton

ssspamqe opened the pull request.

ssspamqe committed:
Use environmental variables to pass api key

ssspamqe committed:
Use java 21 with preview features

ssspamqe committed:
Add SlashCommand classes

ssspamqe committed:
Add tests for HelpSlashCommand

ssspamqe committed:
Add tests for StartSlashCommand

ssspamqe committed:
Fix typo in HelpSlashCommandsTest

ssspamqe committed:
Add repository dependency to slash commands

/list, /track and /untrack must have dependency of repository

ssspamqe committed:
Change parameters of SlashCommand#getSendMessageRequest to just Message

Some additional information can be required while generating response message (message text, userId, user name, etc)

ssspamqe committed:
Rewrite Help and Start SlashCommands tests

Change test according to new parameters (just Message instance) of getSendMessageReqeust

ssspamqe committed:
Add tests for ListSlashCommand

ssspamqe committed:
Rename getSendMessageRequest to getResponse

These method already had type "SendMessage"

ssspamqe committed:
Rename getSendMessageRequest to getSImpleResponse

These method already had type "SendMessage", so there is no need in specifications
Simple -> as command might be parameterized, we need to make a "simple" response to user to ask for additional details/paramers

ssspamqe committed:
Add ParameterizedSlashCommand interface

For handling parameters  of commands

ssspamqe committed:
Merge branch 'hw1-add-bot-skeleton' of https://github.com/ssspamqe/tinkoff-backend-2024 into hw1-add-bot-skeleton

ssspamqe committed:
Refactor command interfaces

Make responseMethods return String because i feel that it will be better in such way

ssspamqe committed:
Refactor SlashCommand interface

Delete Message parameters from getSimpleResponse()
Delete test of simple reponses, because they are just tests of getters

ssspamqe committed:
Make ListSlashCommand implement ParameterizedSlashCommand

ssspamqe committed:
Rewrite tests for ListSlashCommand

Because now it implements ParameterizedSlashCommand

ssspamqe committed:
Add tests for TrackSlashCommand

ssspamqe committed:
Add tests for UntrackSlashCommand

ssspamqe committed:
Rename SlashCommand's method for beter understanding

ssspamqe committed:
Add temporary data storage

Until postgresql  will be connected to the project

ssspamqe committed:
Fix typos in HashMapSubscriptionRepository

ssspamqe committed:
Add tests for HashMapSubscriptionRepository

ssspamqe committed:
Add finding next id for HashMapSubscriptionRepository

ssspamqe committed:
Update tests for HashMapSubscriptionDatabase

ssspamqe committed:
Rename command interfaces and set up sealed

Renamed to make them a bit more clear to understand

ssspamqe committed:
Add controller layer

ssspamqe committed:
Add CommandService

ssspamqe committed:
Add ListSlashCommandImplementation

ssspamqe committed:
Update SlashCommandTests

User getters instead of static strings

ssspamqe committed:
Simplify regex in Subscription

It was hard for user to read old long version

ssspamqe committed:
Fix tests for TrackSashCommand

ssspamqe committed:
Add implementation for TrackSlashCommand

ssspamqe committed:
Fix typo in TrackSlashCommand

ssspamqe committed:
Add implementation for UntrackSlashCommand

ssspamqe committed:
Fix tests for UntrackSlashCommand

ssspamqe committed:
Set up replyMarkup in CommandService

ssspamqe committed:
Add exception handling in Controller layer

ssspamqe committed:
Fix Class casting to BotCommand[]

ssspamqe committed:
Fix typo in TrackSlashCommand class name

ssspamqe committed:
Add NEED_ADDITIONAL_USER_PARAMETERS filed to SlashCommands

It was needed for chooseing between using ReplyKeyboardRemove() and ForceReply() on the message

ssspamqe committed:
Add exception hiearchy for CommandService

ssspamqe committed:
Use exceptions in CommandController

Exceptions are better to use, than regular constant String

ssspamqe committed:
Add exceptions to CommandService

To make it more simple to handle different situations

ssspamqe committed:
Rename methods in CommandService

To make it simpler to understand parameters' role

ssspamqe committed:
Fix ComandService#isUserParameters

ssspamqe committed:
Fix CommandServce#DefineSlashCommand

Make it to be able to define any command, to be able to use it everywhere

ssspamqe committed:
Add exception handler in CommandService

ssspamqe committed:
Change exception handing in CommandServce

To be able to throw other exceptions further

ssspamqe committed:
Add tests for CommandService

ssspamqe committed:
Fix typo in UntrackSlashCommandTest

ssspamqe committed:
Delete redundant exception

ssspamqe committed:
Fix typos

ssspamqe committed:
Add null check in repository

ssspamqe committed:
Refactor BotController

ssspamqe committed:
Fix checkstyle's suggestions

ssspamqe committed:
Update bot.yml

Runt test with --enable-preview flag

ssspamqe committed:
Revert "Update bot.yml"

This reverts commit 8e50abf318195726f42cf754390b4955438671ee.

ssspamqe committed:
Refactor CommandController

Make cosmetic changes in static strings to make them clear
Make CommandServce#defineSlashCommand to get Command from any string

ssspamqe committed:
Rename constants in SlashCommands

ssspamqe committed:
Simplify CommandService#definceSlashCommand

to increase code readability and make it concise

ssspamqe committed:
Delete CantDefineSlashCommandFromTextException.java

No more need in this exception

ssspamqe committed:
Delete test for CommandService

There is no way of getting this exception anymore

ssspamqe committed:
Fix Checkstyle

ssspamqe committed:
Update bot.yml

ssspamqe committed:
Revert "Update bot.yml"

This reverts commit 2a1125137bef33df2738080e86912881bcf9ff58.

ssspamqe committed:
Refactor tests

To make them clear and concise

ssspamqe committed:
Add async execution

ssspamqe committed:
Add --enable-preview flag

ssspamqe committed:
Add test for TrackSlashCommand

to handle situation when given link was already added to database

ssspamqe committed:
Update TrackSlashCommand.java

To handle situations with repeating links

ssspamqe committed:
Update poms

To enable preview features

ssspamqe committed:
Fix TrackSlashCommand#subscriptionWasPreviouslyAdded

ssspamqe committed:
Update TrackSlashCommandTest.java

Omit additional businesses logic checks

ssspamqe committed:
Update poms

ssspamqe committed:
move slashCommands to services package

because they are service

ssspamqe committed:
Fix typo in LoggingCallback

ssspamqe committed:
add CommandService#isCommand

ssspamqe committed:
add host parsing to ListSlashCommand

ssspamqe committed:
Rename SlashCommand interfaces and methds

to increase concise and readability

ssspamqe committed:
Rename CommandService#defineSlashCommand

new name describes method better

ssspamqe committed:
Delete redundant switch branch in CommandService

since SlashCommand permits only 2 interfaces, which was previously checked

ssspamqe committed:
Rename methods

ssspamqe committed:
Fix checkstyle

Checkstyle was falling after STR with """

venommile commented on bot/src/main/java/edu/java/bot/controllers/TelegramBotController.java:
```diff
@@ -0,0 +1,61 @@
+        initializeBot();
+    }
+
+    private void initializeBot() {
```
Используй  хотя бы @PostContstruct, для такого.
Хотя Я бы пошёл бы дальше -- сделал бы TelegramBot ,бином, а логику execute и т.д. Вынес бы в какой-нибудь ContextRefreshedListener или commandLineRunner

venommile commented on bot/src/main/java/edu/java/bot/data/repositories/HashMapRepository.java:
```diff
@@ -0,0 +1,101 @@
+            return Optional.empty();
+        }
+
+        return Optional.of(object);
```
почему не просто Optional.ofNullable()? и if тогда будет не нужен

ssspamqe replied:
потому что я спал пока писал этот модуль

venommile commented on bot/src/main/java/edu/java/bot/services/CommandService.java:
```diff
@@ -0,0 +1,116 @@
+    }
+
+    private void fillAllCommands(List<SlashCommand> commandList) {
+        allCommands = new HashMap<>();
```
давай сразу выделим память. Это можно сделать через commandList.size()/LOAD_FACTOR, по дефолту LOAD_FACTOR=0.75, или через новый метод HashMap.newHashMap(commandList.size())

venommile commented on bot/src/test/java/edu/java/bot/slashCommands/ListSlashCommandTest.java:
```diff
@@ -0,0 +1,91 @@
+
+    @BeforeEach
+    void init() {
+        MockitoAnnotations.openMocks(this);
```
Сделай просто над классом @ExtendWith(MockitoExtension.class)

ssspamqe committed:
Use @ExtendWith(MockitoExtension.class) in tests

ssspamqe committed:
Remove redundant stubs in tests

ssspamqe committed:
Use Optional.ofNullable() in HashMapReposotory

ssspamqe committed:
Predefine allComands map size in CommandService

To increase performance

ssspamqe committed:
Make telegramBot bean and use Spring's events

ssspamqe committed:
Rename package with BotMessageEventListener

ssspamqe committed:
Use async for botMessageEvents

ssspamqe committed:
Reformat all code in bot package

ssspamqe committed:
Make BotMessageEvent a record

ssspamqe committed:
Add tests for BotMessageEventListener

ssspamqe committed:
Change telegramBot implementation

New implementation fixes cases, when telegramBot starts to listen to telegram updates before spring context refreshed

ssspamqe committed:
Delete redundant tests

 There is no such class already

ssspamqe committed:
Add threadPool to BotInitialiaing

ssspamqe committed:
Fix typo in ListSlashCommand

ssspamqe committed:
Add changelog
